python/cuml/common/array_sparse.py

```diff
@@ -0,0 +1,131 @@
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from cuml.common.import_utils import has_scipy
+
+import cupyx as cpx
+import numpy as np
+
+from cuml.common.memory_utils import with_cupy_rmm
+
+from cuml.common.input_utils import input_to_cuml_array
+
+if has_scipy():
+    import scipy.sparse
+
+
+class SparseCumlArray:
+
+    @with_cupy_rmm
+    def __init__(self, data=None, dtype=None):
+        """
+        SparseCumlArray abstracts sparse array allocations. This will
+        accept either a Scipy or Cupy sparse array and construct CumlArrays
+        out of the underlying index and data arrays. Currently, this class
+        only supports the CSR array format and input in any other sparse
+        format will be converted to CSR.
+
+        Parameters
+        ----------
+
+        data : scipy.sparse.csr_matrix or cupyx.scipy.sparse.csr_matrix
+            A Scipy or Cupy sparse csr_matrix
+        dtype : data-type, optional
+            Any object that can be interpreted as a numpy or cupy data type.
+
+        Attributes
+        ----------
+
+        indptr : CumlArray
+            Compressed row index array
+        indices : CumlArray
+            Column indices array
+        data : CumlArray
+            Data array
+        dtype : dtype
+            Data type of data array
+        shape : tuple of ints
+            Shape of the array
+        nnz : int
+            Number of nonzeros in underlying arrays
```

Comment:
We may also want to set `has_sorted` indices for completeness. 

https://github.com/scipy/scipy/blob/01d8bfb6f239df4ce70c799b9b485b53733c9911/scipy/sparse/csr.py#L59-L60

Unsure about the use case for unsorted indices TBH .

Reply:
@VibhuJawa, actually you raise a good question. KNN is a perfect example use-case for unsorted indices. The column insides and data arrays need to remain sorted by the nearest neighbors to each row index. If they are sorted, any nearest neighbors algorithms accepting a nearest neighbors graph will yield incorrect results because they are relying on positional information.

Reply:
Ahh, Nice, Thanks for giving context about the internals of KNN graph.  It makes complete sense to have it sorted by the nearest neighbors to each row index.

 Agreed adding will help possibly detect this scenario. 👍

Reply:
@VibhuJawa I added the `has_sorted_indices` attribute, if you want to take a look.